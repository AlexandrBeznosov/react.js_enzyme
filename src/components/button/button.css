.button-style {
    display: flex;
    justify-content: center;
    align-items: center;
    flex-direction: column;
    width: 200px;
    height: 50px;
    background-color: orange;
    color: white;
    border-radius: 10px;
    margin: 10px;
    cursor: pointer;
}