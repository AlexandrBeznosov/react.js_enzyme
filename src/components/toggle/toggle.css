.toggle-style {
    margin: 10px;
    cursor: pointer;
}