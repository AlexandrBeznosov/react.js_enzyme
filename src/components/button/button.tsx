import { useCallback } from "react";

import './button.css';

export interface ButtonProps {
    title: string;
    isDisabled: boolean;
    onClick: (message: string) => void;
}

export const Button: React.FC<ButtonProps> = props => {
    const { onClick } = props;

    const handleClick = useCallback(() => {
        if (!props.isDisabled) {
            onClick('Hello!');
        }
    }, [onClick, props.isDisabled]);

    return <div className="button-style" data-test-target="button-element" onClick={handleClick}>{props.title}</div>
}