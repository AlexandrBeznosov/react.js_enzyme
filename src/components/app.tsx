import React from 'react';

import { Button } from './button/button';
import { Toggle } from './toggle/toggle';

const App: React.FC = () => {
    return (<div className="App">
        <Button title='Click me!' onClick={message => window.alert(message)} isDisabled={false} />
        <Toggle title='Toggle me!' onChange={state => window.console.log(String(state))} initialState={true} />
    </div>);
}

export { App };
