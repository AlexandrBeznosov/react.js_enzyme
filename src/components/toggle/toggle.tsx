import { useCallback, useState } from "react";

import './toggle.css';

export interface ToggleProps {
    title: string;
    initialState: boolean;
    onChange: (state: boolean) => void;
}

export const Toggle: React.FC<ToggleProps> = props => {
    const { onChange } = props;

    const [state, setState] = useState(props.initialState);

    const handleChange = useCallback(() => {
        setState(!state);
        onChange(!state);

    }, [onChange, state]);

    return (
        <label className="toggle-style" htmlFor="toggle-id">
            <input type="checkbox" checked={state} id="toggle-id" data-test-target="toggle-element" onChange={handleChange} />
            {props.title}
        </label>
    );
}