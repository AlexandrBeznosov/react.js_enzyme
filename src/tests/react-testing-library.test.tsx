import React from 'react';
import { fireEvent, logRoles } from '@testing-library/dom';
import { render, screen } from '@testing-library/react';
import { App } from '../components/app';
import { Toggle, ToggleProps } from '../components/toggle/toggle';
import { Button, ButtonProps } from '../components/button/button';

describe('React Testing Library do:', () => {

    test('create snapshot of mounted component', () => {
        // render app
        render(<App />);

        // check the snapshot
        expect(screen).toMatchSnapshot;
    });

    test('find child by component', () => {
        // render App
        render(<App />);
        

        // find the component
        const childElement = screen.getByText(/click/i);

        // check if it exists
        expect(childElement).toBeInTheDocument();
    });

    test('find child by attribute', () => {
        // render App
        const {container} = render(<App/>);

        // find the component 
        const childElement = container.querySelector('[data-test-target="button-element"]'); // had to use query selector because getByTestId doesn't work since that function requires a data attribute named data-testid.

        // check if it exists
        expect(childElement).toBeInTheDocument();
    });

    test('find child by class name', () => {
        // render component
        const props: ToggleProps = {
            initialState: true,
            title: 'Dummy toggle title',
            onChange: () => null,
        };
        const {container} = render(<Toggle {...props} />);
        

        // find the component
        const input = container.querySelector('.toggle-style');

        // check if it exists
        expect(input).toBeInTheDocument();
    });

    test('get child attribute', () => {
        // render app
        render(<App/>);

        // find the component and it's attribute
        const input = screen.getByLabelText(/toggle/i);

        // simulate input
        fireEvent.click(input);

        // check the attribute value 
        expect(input).not.toBeChecked(); // since the initial state of the checkbox is checked, im asserting wether or not it is unchecked.
    });

    test('set component property', () => {
        // create a dummy button
        const props: ButtonProps = {
            title: "Dummy button title",
            isDisabled: false,
            onClick: jest.fn(),
        }

        // render and disable the button
        const {rerender} = render(<Button {...props}/>);
        rerender(<Button {...props} isDisabled={true} />)

        // find the button
        const button = screen.getByText(/dummy/i);

        // simulate a click
        fireEvent.click(button);
        
        // check if the button is clickable
        expect(props.onClick).toBeCalledTimes(0);
    });

    test('callback test', () => {
        // create a dummy button
        const props: ButtonProps = {
            title: "Dummy button props",
            isDisabled: false,
            onClick: jest.fn(),
        }

        // render the button
        render(<Button {...props}/>);

        // find the button
        const button = screen.getByText(/dummy/i);

        // simulate a click
        fireEvent.click(button);

        // check if the button is clicked
        expect(props.onClick).toBeCalledTimes(1);
    });
});
