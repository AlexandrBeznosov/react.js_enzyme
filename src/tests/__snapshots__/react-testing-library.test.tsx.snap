// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React Testing Library do: create snapshot of mounted component 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div
      class="button-style"
      data-test-target="button-element"
    >
      Click me!
    </div>
    <label
      class="toggle-style"
      for="toggle-id"
    >
      <input
        checked=""
        data-test-target="toggle-element"
        id="toggle-id"
        type="checkbox"
      />
      Toggle me!
    </label>
  </div>
</DocumentFragment>
`;
